---
title: "CASC"
author: "Raffaele A Calogero"
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## CASC Classification Analysis of Single Cell Sequencing Data

Genome-wide single-cell measurements such as transcriptome sequencing enable the characterization of cellular composition as well as functional variation in homogenic/heterogenic cell populations. An important step in the single-cell transcriptome analysis is to group cells that belong to the same sub-type based on gene expression patterns [1-3]. Critical issues in cell clustering are:

- cluster stability

- feature selection, i.e. the identification genes playing the major role in cluster formation. 

To address the above issues, we have developed CASC, a tool implemented in a docker container, that uses as core application to detect cell clusters the “kernel based similarity learning” [4] and allows: 

- identification of the optimal number of clusters for cell partitioning using “silhouette method”. 

- The evaluation of clusters stability, measuring the permanence of a cell in a cluster upon random removal of subsets of cells. 

-Feature selection via “nearest shrunken centroid method” [5], applied to the gene Index Of Dispersion [6]. 

CASC was tested on previously published data sets [1-3]. CASC feature selection procedure efficiently allows the identification of the subpopulation of genes playing the main role in discriminating between cell subpopulations.


## Introduction

**CASC** is controlled by a set of functions embedded in `r Githubpkg("kendomaniac/docker4seq")` package,
which was developed to facilitate the use of computing demanding applications in the field of NGS data analysis.


This approach provides multiple advantages: 

- user does not need to install all the software on its local server

- results generated by different containers can be organized in pipelines

- reproducible research is guarantee by the possibility of sharing the docker containers used for the analysis 

## Requirements
The minimal hardware requirements are a 4 core 64 bits linux computer, 32 Gb RAM, one SSD 250GB, with a folder with read/write permission for any users (chmod 777), and [docker](https://www.docker.com/) installed.

[**docker4seq**](https://github.com/kendomaniac/docker4seq) can fit ideally in the [NUC6I7KYK](https://www.intel.com/content/www/us/en/products/boards-kits/nuc/kits/nuc6i7kyk.html) equipped with Kingston Technology HyperX Impact 32GB Kit (2x16GB), 2133MHz DDR4 CL13 260-Pin SODIMM and Samsung 850 EVO - 250GB - M.2 SATA III Internal SSD.
A ready to go fedora installation USB stick, with everything installed to be used with the above hardware configuration, can be requested to: raffaele[dot]calogero[at]unito[dot]it

**IMPORTANT** The first time *docker4seq* is installed the **downloadContainers** needs to be executed  to download to the local repository the containers that are needed for the use of *docker4seq*

```{r, echo=TRUE, eval=FALSE}
downloadContainers(group="docker")
```




## References

[1] Usoskin et al. Nat. Neurosci. 2014, 18:145–153

[2] Pollen et al. Nat. Biotechnol. 2014, 32:1–37

[3] Kolodziejczyk et al. Cell Stem Cell 2015, 17:471–485

[4] Wang et al. Nat Methods. 2017 14:414-416

[5] Tibshirani et al. PNAS 2002, 99:6567-6572

[6] Diaz et al. Bioinformatics 2016, 32: 2219-2220

