---
title: "An Introduction to docker4seq package"
author: "Raffaele A Calogero"
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction
The `r Githubpkg("kendomaniac/docker4seq")` package was developed to facilitate the use computing demamding steps in NGS data analysis.

The `r Githubpkg("kendomaniac/docker4seq")` package is the use of docker contaniers that embed demanding computing task as short reads mapping and counting. 
This approach provides multiple advantages: 
*user does not need to install all the software on its local server
*results generated by different containers can be organized in pipelines
*reproducibility is guarantee by the possibility of sharing the docker containers used for the analysis 

## Requirements
The minimal hardware requirements are a 4 core 64 bits linux computer, 32 Gb RAM, one SSD 250GB, with a folder with read/write permission for any users (chmod 777), and [docker](https://www.docker.com/) installed.

## Available workflows
At the present time are available the following workflows:

* mirnaCounts executes the workflow described in Cordero et al. PLoS One. 2012;7(2):e31630, which embeds the following steps:
    + trimming adapters with [cutadapt](http://cutadapt.readthedocs.io/en/stable/guide.html)
    + miRNAs mapping on [mirbase](http://www.mirbase.org/) hairpins using [SHRiMP](http://compbio.cs.toronto.edu/shrimp/)
    + quantification of mature miRNAs.
* rnaseqCounts, which allows:
    + adapter trimming with [skewer](https://github.com/relipmoc/skewer)
    + mapping with [STAR](https://github.com/alexdobin/STAR)
    + counting genes and isoforms with [RSEM](http://deweylab.github.io/RSEM/)
    + ENSEMBL gene annotation.
* chipseq, which allows:
    + adapter trimming with [skewer](https://github.com/relipmoc/skewer)
    + mapping with [BWA](http://bio-bwa.sourceforge.net/bwa.shtml)
    + peak calling using either MACS v 1.4 or SICER v 1.1
    + connection of the peak to the nearest gene available in the chosen UCSC genome release,
    + full annotation of the nearest gene.
    
## chipseq workflow

The chipseq workflow can be run using 4SeqGUI graphical interface:
![ChIPseq workflow](/Users/raffaelecalogero/Dropbox/data/docker/docker4seq_devel/docker4seq_devel/inst/img/chipseq0.jpeg)


The ChIPseq is made of two main steps:

- Creating a genome index for BWA (see end of this paragraph)

- Running MACS or SICER analysis

All the parameters needed to run MACS or SICER can be setup using 4SeqGUI

![MACS and SICER analysis](/Users/raffaelecalogero/Dropbox/data/docker/docker4seq_devel/docker4seq_devel/inst/img/chipseq3.jpeg)


A detailed description of the parameters is given below.

The chipseq workflow can be also executed using R and it is completely embedded in a unique function:

```{r, echo=TRUE, eval=FALSE}

library(docker4seq)

chipseqCounts(group = c("sudo", "docker"), output.folder = getwd(),
  mock.folder, test.folder, scratch.folder,
  adapter5 = "AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT",
  adapter3 = "AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT",
  threads = 8, min.length = 30, genome.folder,
  mock.id = "igg", test.id = "tf", genome, read.size = 50,
  tool = "macs", macs.min.mfold = 10, macs.max.mfold = 30,
  macs.pval = "1e-5", sicer.wsize = 200, sicer.gsize = 200,
  sicer.fdr = 0.1, tss.distance = 0, max.upstream.distance = 10000,
  remove.duplicates = "N")
```

Specifically user needs to create three folders:

    + mock.folder, where the fastq.gz file for the control sample is located. For control sample we refer to ChIP with IgG only or input DNA.
    + test.folder, where the fastq.gz file for the ChIP of the sample to be analysed.
    + output.folder, where the R script embedding the above script is located.

The **scratch.folder** can be the same as the **output.folder**. However, if the system in use has a high speed disk for temporary calculation, e.g. a SSD disk, the location of the scratch.folder on the SSD will reduce significantly the computing time.

User needs to provide also the sequence of the sequencing adapters, **adapter5** and **adapter3** parameters. In case Illumina platform is used the adapters sequences can be easily recovered [here](https://support.illumina.com/content/dam/illumina-support/documents/documentation/chemistry_documentation/experiment-design/illumina-adapter-sequences_1000000002694-01.pdf). 

**Threads** indicates the max number of cores used by *skewer* and *bwa*, all the other steps are done on a single core.
The **min.length** refers to the minimal length that a reads should have after adapters trimming. Since today the average read length for a ChIP experiment is 50 or 75 nts would be better to bring to 40 nts the min.length parameter to increase the precision in assigning the correct position on the genome.

The **genome.folder** parameter refers to the location of the genomic index generated by bwa using the *docker4seq* function **bwaIndexUcsc**.
The generation of the genome index for ChIP experiment is very simple and it is highlited at the end of this paragraph.

**mock.id** and **test.id** identify the type of sample and are assigned to the ID parameter in the RG field of the bam file.

**genome** is the parameter referring to the annotation used to associate ChIP peaks to genes. In the present implemetation are available hg38, hg19 for human and mm10 and mm9 for mouse annotations.

**read.size** is a parameter requested by MACS and SICER for their analysis. 
**macs.min.mfold**, **macs.max.mfold**, **macs.pval**  are the deafult parameters requested for  peaks definition for more info please refer to the documetation of MACS 1.4.
**sicer.wsize**, **sicer.gsize**, **sicer.fdr** are the deafult parameters requested for  peaks definition for more info please refer to the documetation ofSICER 1.1.
**Important**: The optimal value for **sicer.gsize** in case of H3K4Me3 ChIP is 200 and in case of ChIP H3K27Me3 is 600.   

**tss.distance** and **max.upstream.distance** are parameters required by ChIPseqAnno, which is the bioconducto package used to assign the peaks to specific genes. Specifically max.upstream.distance refers to the max distance in nts that allow to associate a peak to a gene.

**remove.duplicates** is the parameter that indicates if duplicates need to be removed or not. It has two options: **N** duplicates are not removed, **Y** duplicates are removed.




### Creating a BWA index file for Chipseq:

```{r,  echo=TRUE, eval=FALSE}
bwaIndexUcsc(group="sudo",genome.folder="/sto2/data/scratch/hg19_bwa", uscs.urlgenome=
"http://hgdownload.cse.ucsc.edu/goldenPath/mm10/bigZips/chromFa.tar.gz",
gatk=FALSE)
```

In brief, **bwaIndexUcsc** uses UCSC genomic data. User has to provide the URL (**uscs.urlgenome**) for the file chromFa.tar.gz related to the organism of interest and the path to the folder where the index will be generated (**genome.folder**). The parameter **gatk** has to be set to FALSE because is not used for a genomic index used for ChIPseq.

The index can be easily created using the graphical interface:

![Creating a BWA index with Genome indexing BWA](/Users/raffaelecalogero/Dropbox/data/docker/docker4seq_devel/docker4seq_devel/inst/img/chipseq1.jpeg)



