#' @title Running RNAseq counting workflow for a single sample
#' @description This function executes a set of docker containers allowing the generation of gene and isoforms counts for a single sample.
#' #params skewer
#' @param group, a character string. Two options: \code{"sudo"} or \code{"docker"}, depending to which group the user belongs
#' @param fastq.folder, a character string indicating where gzip fastq files are located
#' @param scratch.folder, a character string indicating the scratch folder where docker container will be mounted
#' @param adapter5, a character string indicating the fwd adapter
#' @param adapter3, a character string indicating the rev adapter
#' @param seq.type, a character string indicating the type of reads to be generated by the sequencer. Two options: \code{"se"} or \code{"pe"} respectively for single end and pair end sequencing.
#' @param threads, a number indicating the number of cores to be used from the application
#' @param min.length, a number indicating minimal length required to return a trimmed read
#' #params rsemstar
#' @param genome.folder, a character string indicating the folder where the indexed reference genome for bwa is located. IMPORTANT the present function only suport genomic indexes made using ensembl genom and the corresponding gtf
#' @param strandness, a character string indicating the type ofsequencing protocol used for the analysis. Three options: \code{"none"}, \code{"forward"}, \code{"reverse"} respectively for non strand selection, forward for Illumina strandness protocols, reverse for ACCESS Illumina protocol
#' #params rsemanno
#' @param org, a character string indicating the genome assembly used for mapping and counting with \code{"rsemstar"} function
#' @param truncating.expected.counts, a boolean logical variable indicating if the expected counts calculated by RSEM need to be converted in integer to be compliant with differnetial expression Bioconductor packages as DESeq2. Default is FALSE
#' @param annotation.type, a character string. Two options: \code{"rsemUCSC"} or \code{"rsemENSEMBL"}, depending to which database was used in rsemstar for counting: UCSC or ENSEMBL

#' @return Returns the output of skewer, rsemstar, rsemanno
#' @examples
#'\dontrun{
#'     wget http://130.192.119.59/public/test_R1.fastq.gz
#'     wget http://130.192.119.59/public/test_R2.fastq.gz
#'     rnaseqCounts(group="sudo",fastq.folder=getwd(), scratch.folder="/data/scratch",
#'     adapter5="AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT",
#'     adapter3="AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT",
#'     seq.type="pe", threads=10,  min.length=40,
#'     genome.folder="/data/scratch/hg38star", strandness="none",
#'     org="hg38", truncating.expected.counts=FALSE, annotation.type="rsemENSEMBL")
#' }
#' @export
rnaseqCounts<- function( group="sudo",fastq.folder=getwd(), scratch.folder="/data/scratch", threads=4,
adapter5="AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT",
adapter3="AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT",
seq.type="pe",   min.length=40,
genome.folder="/data/scratch/hg38star", strandness="none",
org="hg38", truncating.expected.counts=FALSE, annotation.type="rsemENSEMBL"){
  #trimming adapter
  skewer(group=group,fastq.folder=fastq.folder, scratch.folder=scratch.folder,adapter5=adapter5, adapter3=adapter3, seq.type=seq.type, threads=threads,  min.length=min.length)
  #running rsemstar
  rsemstar(group=group,fastq.folder=fastq.folder, scratch.folder=scratch.folder, genome.folder=genome.folder, seq.type=seq.type, strandness=strandness,threads=threads)
  #rumming annotation
  rsemanno(group=group,rsem.folder=fastq.folder, scratch.folder=scratch.folder, org=org, truncating.expected.counts=truncating.expected.counts, annotation.type=annotation.type, protein.anno=FALSE)
}


